### PR TITLE
Login

### PR DESCRIPTION
**Pull request by barbarakos**
Added LoginVM, LoginUseCase, LoginDatasource and LoginClient.

### CODE REVIEW

File: QuizApp/Login/LoginClient.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+class LoginClient {
+
+    let baseURL = "https://five-ios-quiz-app.herokuapp.com/"
+    let loginPath = "v1/login"
+
+    func login(password: String, username: String) async throws -> LoginResponseModel {
+        guard let URL = URL(string: "\(baseURL)\(loginPath)") else {
+            throw RequestError.invalidURL
+        }
+        var URLrequest = URLRequest(url: URL)
```
Comment (marinzg) [inline QuizApp/Login/LoginClient.swift:21]: New line after `guard`.

File: QuizApp/Login/LoginClient.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+enum RequestError: Error {
+
+    case clientError
+    case serverError
+    case dataError
+    case invalidURL
+    case unknown
+
+}
```
Comment (marinzg) [inline QuizApp/Login/LoginClient.swift]: Put models into a separate file.

File: QuizApp/Login/LoginDatasource.swift [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+class LoginDatasource {
+
+    private var loginClient = LoginClient()
+
+    func login(username: String, password: String) async throws -> LoginResponseModel {
+        let response: LoginResponseModel = try await loginClient.login(password: password, username: username)
+        return response
```
Comment (marinzg) [inline QuizApp/Login/LoginDatasource.swift]: Can't you just return from client?
```swift
func login(username: String, password: String) async throws -> LoginResponseModel {
    return try await loginClient.login(password: password, username: username)
}
```

File: QuizApp/Login/LoginDatasource.swift [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+class LoginDatasource {
```
Comment (marinzg) [inline QuizApp/Login/LoginDatasource.swift:9]: For each implementation, you need to have a protocol.
This goes for usecases, repositories, datasources and clients.

File: QuizApp/Login/LoginViewController.swift [hunk 5]
```diff
@@ -16,6 +17,7 @@ class LoginViewController: UIViewController {
         self.init()
 
         self.router = router
+        loginViewModel = LoginViewModel(router: router)
```
Comment (marinzg) [inline QuizApp/Login/LoginViewController.swift]: You shouldn't create view model in the view controller, it should all be done in the router:
```swift
// Router
let viewModel = MyViewModel(router: Router(), dependency: Dependency())
let viewController = MyViewController(viewModel: viewModel)
// display viewController

// MyViewController
...
init(viewModel: MyViewModel) {
    self.viewModel = viewModel
}


// My ViewModel
...
init(router: Router, dependency: Dependency) {
    self.router = router
    self.dependency = dependency
}
```

File: QuizApp/Login/LoginViewController.swift [hunk 6]
```diff
@@ -30,6 +32,7 @@ class LoginViewController: UIViewController {
     }
 
     @objc func handleLogIn() {
+        loginViewModel.login(username: emailTextField.text!, password: passwordTextField.text!)
```
Comment (marinzg) [inline QuizApp/Login/LoginViewController.swift]: I know you're sure there's going to be something here, but it would be better to unwrap it with guard.

**Review by marinzg — CHANGES_REQUESTED** [review]

File: QuizApp/Login/LoginDatasource.swift [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+protocol LoginDatasourceProtocol {
+
+    func login(username: String, password: String) async throws -> LoginResponseModel
+    
+}
+
+class LoginDatasource {
+
+    private var loginClient = LoginClient()
```
Comment (marinzg) [inline QuizApp/Login/LoginDatasource.swift:11]: Same thing as for view controller.
All classes should have their dependencies injected through init.

**Review by marinzg — CHANGES_REQUESTED** [review]

**Review by marinzg — APPROVED** [review]

File: QuizApp/Login/LoginDatasource.swift (hunk 4 above)
Comment (barbarakos) [reply QuizApp/Login/LoginDatasource.swift:9]: https://github.com/barbarakos/QuizApp/pull/8

**Pull request merged**